### PR TITLE
Implemented support for multiple types.db files in config.toml.

### LINKED ISSUE #1110: Support for multiple types.db files in config.toml
Appears influxDB currently has support for just 1 typesdb file.  Those who maintain custom types.db files for their collectd implementations will have to merge types.db's into 1 file if they use "Type" from both.

Reproduce:
  [input_plugins.collectd]
  enabled = true
  port = 25826
  database = "collectd"
  typesdb = "/usr/share/collectd/types.db.custom"
  typesdb = "/usr/share/collectd/types.db"

/opt/influxdb/current/influxdb -config /opt/influxdb/shared/config.toml
Couldn't parse configuration file: /opt/influxdb/shared/config.toml
[11/10/14 15:13:00] [INFO] Loading configuration file /opt/influxdb/shared/config.toml
Near line 57, key 'input_plugins.collectd.typesdb': Key 'input_plugins.collectd.typesdb' has already been defined.

or similarly (using space or , as separator)
 [input_plugins.collectd]
  enabled = true
  port = 25826
  database = "collectd"
  typesdb = "/usr/share/collectd/types.db.custom" "/usr/share/collectd/types.db"

/opt/influxdb/current/influxdb -config /opt/influxdb/shared/config.toml
Couldn't parse configuration file: /opt/influxdb/shared/config.toml
[11/10/14 15:14:22] [INFO] Loading configuration file /opt/influxdb/shared/config.toml
Near line 56, key 'input_plugins.collectd': Near line 56: Expected a top-level item to end with a new line, comment or EOF, but got '"' instead.

Could we support multiple to align with collectd?

### PR DESCRIPTION
**Pull request by pires**
- [ ] CHANGELOG.md updated
- [x] Rebased/mergable
- [x] Tests pass
- [x] Sign CLA (if not already signed)

Fixes #1110

### CODE REVIEW

**Comment by otoolep** [comment]
Unfortunately this is a breaking change. Existing collectd configuration will not work with this code.

Hmmm, need to think about this.

**Comment by otoolep** [comment]
Instead, supporting a directory where multiple types files could be imported from would not be a breaking change. The path for that could be added as a new config option, and would not be a breaking change. And if that directory is non-zero, then the types files are loaded from there (every file in the directory is imported).

@pires ?

**Comment by pires** [comment]
@otoolep yes, I agree.

**Comment by pires** [comment]
@otoolep done. I have decided to support both `typesdb` for single file (current behavior) and `typesdb-dirs` for multiple directories.
However, if any of these directories include a file that can't be parsed, the entire operation will fail. Comments on this?

File: services/collectd/config_test.go [hunk 1]
```diff
@@ -28,5 +29,7 @@ typesdb = "yyy"
 		t.Fatalf("unexpected database: %s", c.Database)
 	} else if c.TypesDB != "yyy" {
 		t.Fatalf("unexpected types db: %s", c.TypesDB)
+	} else if len(c.TypesDBDirs) != 2 {
```
Comment (otoolep) [inline services/collectd/config_test.go]: This test should check that the actual paths are as expected.

**Comment by pires** [comment]
@otoolep tests are failing but none related to this PR, am I right?

```
--- FAIL: TestServer_Query_Aggregates_FloatMany (10.06s)
	server_test.go:2544: distinct as call - float: unexpected results
		query:  SELECT DISTINCT(value) FROM floatmany
		exp:    {"results":[{"series":[{"name":"floatmany","columns":["time","distinct"],"values":[["1970-01-01T00:00:00Z",[2,4,5,7,9]]]}]}]}
		actual: {"results":[{"series":[{"name":"floatmany","columns":["time","distinct"],"values":[["1970-01-01T00:00:00Z",null]]}]}]}
```

File: services/collectd/config_test.go (hunk 1 above)
Comment (pires) [inline services/collectd/config_test.go]: Done.

**Comment by otoolep** [comment]
@pires -- yeah those test failures are something else we're working on.

**Comment by pires** [comment]
@otoolep fixed the test nit. Anything else?

**Comment by otoolep** [comment]
This looks sane to me, though I need to study the gollectd code to ensure what we are doing to merge the files is actually valid. I'm sure you've done this, obviously we need peer review.

Thanks @pires

**Comment by pires** [comment]
@otoolep I have a couple checks for that on both tests I added. See https://github.com/influxdb/influxdb/pull/4740/files#diff-1706d1356c016e994827ad3d884a501aR242 onwards.

* First test-case should load a specific file with doesn't include `custom-metric`.
* Second test-case should load both files in directory and have `custom-metric`.
* Both test-cases should have all the keys included in the specific file loaded in the first test-case, but I chose to validate only `absolute`.

Do you think I should add more tests? If so, can you provide more details?

Thank you!

File: services/collectd/config_test.go (hunk 1 above)
Comment (kimor79) [inline services/collectd/config_test.go]: There are often other files in the default types.db directory (default postgres queries, java files, ...)

**Comment by pires** [comment]
@kimor79 the user should take care of that, if and only if directories are configured. 

@otoolep anything missing?

**Comment by otoolep** [comment]
The code looks good, but I need to study it more closely and I may also need to actually test with a collectd setup. It might be a little while before we merge this.

**Comment by pires** [comment]
@otoolep thank you for the feedback. Let me know if there's anything we can do to improve it.

**Comment by otoolep** [comment]
@pires -- can you actually perform some system test on this? Have collectd send the data to your system, and ensure it actually works the way you expect?

**Comment by pires** [comment]
@AndrejKorshikov want to try what @otoolep as asked for?

**Comment by AndrejKorshikov** [comment]
@pires -- yes, I'll try it in a couple of days, with my collectd setup getting info from UPSes and cooling systems.

**Comment by otoolep** [comment]
OK, thanks for any help you can provide. We must be sure not to break existing collectd functionality, as well be sure that the new functionality works, at a system-test level.

**Comment by pires** [comment]
@AndrejKorshikov eagerly awaiting on your feedback.

**Comment by AndrejKorshikov** [comment]
@pires sorry for delay, I have no access to my collectd setup at this moment. I hope I'll be able to perform tests this week.

**Comment by xaniasd** [comment]
hey, any updates on this?

**Comment by pires** [comment]
@xaniasd we need system tests to validate the PR behavior.

**Comment by xaniasd** [comment]
@pires hi, glad to help if possible, what kind of system tests are needed? the test cases described above can be executed with a single VM, unless I'm missing something. 

Another remark, if I may: collectd configures the types.db files in a space-separated list of strings, like so:
```TypesDB "/usr/share/collectd/types.db" "/etc/collectd/my_types.db"```
wouldn't this work here too? and I believe it's not a breaking change like this (@otoolep?)

**Comment by joelegasse** [comment]
I'm a little confused when reading the code. The added test (and other comments) refer to the `*.db` names as files, but they are processed as directories in the code (and the name is `typesdb-dirs`).

Is the goal to list multiple directories where every file found in them is treated as a typesdb file to be proccessed? Or is the goal to explicitly list additional typesdb files that should be processed?

@kimor79 made the comment that other files might exist alongside custom `types.db` files, and @xaniasd mentioned that the `TypesDB` config option in collectd is a list of files, not directories.

Given all that, I would think the new config option should be `typesdb-extra` and be a list of *files* to be processed, not directories. The current PR code overrides existing entries with the extra files. Is that the desired behavior? Or should the extra files only add new entries, leaving existing entries untouched.

I'm not familiar with collectd, so I'm not sure if overriding defaults or just adding custom types is more common/desired. Either way, the behavior should at least be documented in a comment somewhere (probably the new config field), and ideally there would be a test with multiple files to ensure the code matches the expected behavior (override or otherwise).

Also, @xaniasd, we can't use space separated strings because InfluxDB uses [toml](https://github.com/toml-lang/toml) config files, which treat strings and arrays of strings differently.

**Comment by mengelmann** [comment]
@joelegasse we use collectd quite extensively, so to clarify:
* I think that it's more common to add custom types (for custom metrics) than overriding defaults (web interfaces for collectd use predefined/default collectd settings so changing them is not useful).
* Syntax for TypesDB in collectd configuration is space separated string of paths (as @xaniasd mentioned):
```
TypesDB File [File ...]
Set one or more files that contain the data-set descriptions.
```
(from https://collectd.org/documentation/manpages/collectd.conf.5.shtml#global_options)

**Comment by jsternberg** [comment]
Closing this in favor of #6908 since that one isn't a breaking change.

**Pull request closed without merging**